pghoard/rohmu/object_storage/s3.py

```diff
@@ -62,152 +77,245 @@ def get_metadata_for_key(self, key):
-            if metadata:
-                for k, v in metadata.items():
-                    s3key.set_metadata(k, v)
-            s3key.set_contents_from_filename(filepath, replace=True,
-                                             encrypt_key=self.encrypted)
-        else:
-            start_of_multipart_upload = time.monotonic()
-            chunks = math.ceil(size / self.multipart_chunk_size)
-            self.log.debug("Starting to upload multipart file: %r, size: %r, chunks: %d",
-                           key, size, chunks)
-            mp = self.bucket.initiate_multipart_upload(key, metadata=metadata,
-                                                       encrypt_key=self.encrypted)
-
-            with open(filepath, "rb") as fp:
-                part_num = 0
-                while fp.tell() < size:
-                    part_num += 1
-                    start_time = time.monotonic()
-                    self._store_multipart_upload(mp, fp, part_num, filepath)
-                    self.log.info("Upload of part: %r/%r of %r, part size: %r took: %.2fs",
-                                  part_num, chunks, filepath, self.multipart_chunk_size,
-                                  time.monotonic() - start_time)
-            if len(list(mp)) == chunks:
-                self.log.info("Multipart upload of %r, size: %r, took: %.2fs, now completing multipart",
-                              filepath, size, time.monotonic() - start_of_multipart_upload)
-                mp.complete_upload()
-            else:
-                err = "Multipart upload of {!r} does not match expected chunk list".format(key)
-                self.log.error(err)
-                mp.cancel_upload()
-                raise StorageError(err)
+            with open(filepath, "rb") as fh:
+                data = fh.read()
+                self.store_file_from_memory(key, data, metadata)
+            return
+
+        key = self.format_key_for_backend(key)
+
+        start_of_multipart_upload = time.monotonic()
+        chunks = math.ceil(size / self.multipart_chunk_size)
+        self.log.debug("Starting to upload multipart file: %r, size: %r, chunks: %d",
+                       filepath, size, chunks)
+
+        parts = []
+        part_number = 1
+
+        extra_args = {}
+        if metadata:
+            extra_args["Metadata"] = self.sanitize_metadata(metadata)
+        if self.encrypted:
+            extra_args["ServerSideEncryption"] = "AES256"
+
+        try:
+            response = self.s3_client.create_multipart_upload(
+                Bucket=self.bucket_name,
+                Key=key,
+                **extra_args,
+            )
+        except botocore.exceptions.ClientError as ex:
+            raise StorageError("Failed to initiate multipart upload for {}".format(key)) from ex
+
+        mp_id = response["UploadId"]
+
+        with open(filepath, "rb") as fp:
+            while True:
+                data = fp.read(self.multipart_chunk_size)
```

Comment:
maybe we can afford it, but this changes previous behavior where we didn't have to read the whole 100mb chunk in memory to upload it.

Reply:
Botocore contains a bit of logic on e.g. rewinding stream with seek(0) on redirects, etc. I'm not sure these matter in the actual call flows, but I'd rather keep the explicit control here. I've lowered the multipart size slightly to 64MB.